### PR TITLE
#42: PHP 8.1 compatibility fixes

### PR DESCRIPTION
**Pull request by michalbiarda**
- PHP 8.1 compatibility fixes
- unit tests updates
- codebase cleanup with PHPCS
- improved error handling
- several minor fixes

### CODE REVIEW

File: src/Pipeline/Span/Primary.php [hunk 1]
```diff
@@ -27,29 +13,35 @@
 use Driver\Pipeline\Transport\Status;
 use Driver\Pipeline\Transport\TransportInterface;
 use Driver\System\YamlFormatter;
-use Haystack\HArray;
-use Symfony\Component\Console\Output\OutputInterface;
+
+use function array_filter;
```
Comment (prince-swiftotter) [inline src/Pipeline/Span/Primary.php:17]: @michalbiarda Should we import PHP functions here?

Comment (michalbiarda) [reply src/Pipeline/Span/Primary.php:17]: Actually, importing PHP functions (or using `\function_name` inline) is a good practice. That way you are always sure that you're using function from correct namespace (global in that case).

File: src/Pipeline/Span/Primary.php [hunk 2]
```diff
@@ -109,17 +105,21 @@ private function sortStages(array $stages)
...
-    private function mapToStageObjects(array $stages)
+    /**
+     * @return StageInterface[] $stages
+     */
+    // phpcs:ignore SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingTraversableTypeHintSpecification
+    private function mapToStageObjects(array $stages): array
     {
-        return array_filter(array_map(function($stage, $name) {
+        return array_filter(array_map(function ($stage, $name) {
             if (isset($stage['actions'])) {
```
Comment (prince-swiftotter) [inline src/Pipeline/Span/Primary.php:123]: @michalbiarda We can optimize this if else using ternary here into single line to return value

File: src/Pipeline/Span/Primary.php (hunk 1 above)
Comment (prince-swiftotter) [reply src/Pipeline/Span/Primary.php:17]: Thanks for clarifying.

**Pull request merged**